training/dataset.py

```diff
@@ -14,16 +19,88 @@ def __len__(self):
     def shuffle(self):
         np.random.shuffle(self.data)
 
-    def iterbatches(self, batch_size):
+    def iterbatches(self, batch_size: int) -> List[Any]:
         batch = 0
         for _ in range(self.num_batches(batch_size)):
             start = batch*batch_size
             end = min((batch+1)*batch_size, len(self.data))
             yield self.data[start:end]
             batch += 1
 
-    def split(self, frac):
+    def split(self, frac: float) -> Tuple["Dataset", "Dataset"]:
+        """Split dataset into two, using a fraction.
+
+
+        Parameters
+        ----------
+
+        frac: float
+            The proportion of the dataset to have to split on.
+
+        Returns
+        -------
+        tuple(Dataset, Dataset)
+            The split datasets, the first dataset containing the fractional percentage
+            and the second with the leftover.
+
+        """
+        if frac > 1.0 or frac < 0.0:
+            raise ValueError("frac must be between 0 and 1")
         split_idx = int(len(self.data)*frac)
         train_dataset = self.data[:split_idx]
         test_dataset = self.data[split_idx:]
         return Dataset(train_dataset), Dataset(test_dataset)
+
+    def random_split(self, frac: float) -> Tuple["Dataset", "Dataset"]:
+        """Shuffle data and split into two datasets, using a fraction.
+
+
+        Parameters
+        ----------
+
+        frac: float
+            The proportion of the dataset to have to split on.
+
+        Returns
+        -------
+        tuple(Dataset, Dataset)
+            The split datasets, the first dataset containing the fractional percentage
+            and the second with the leftover
+
+        """
+        inds = np.arange(len(self))
+        np.random.shuffle(inds)
+        split_idx = int(len(self.data)*frac)
+        return self.indice_split(inds[:split_idx], inds[split_idx:])
+
+    def indice_split(self, left: List[int], right: List[int]) -> Tuple["Dataset", "Dataset"]:
+        """Split dataset into two, providing the indices. Used when the data needs to
+        be split in more specific ways that the Dataset class does not provide out of the box.
+
+
+        Parameters
+        ----------
+
+        left: array of integers
+            Indices of dataset to construct first dataset from.
+
+        right: array of integers
+            Indices of dataset to construct second dataset from.
+
+        Returns
+        -------
+        tuple(Dataset, Dataset)
+            The split datasets, the first dataset containing the objects at the indices
+            provided by left and the second with the right indices.
+
+        """
+        left = set(left)
+        right = set(right)
+        indices = set(range(len(self)))
+        if len(left.intersection(right)) > 0:
+            raise ValueError("Left and right indices contain overlap")
+        if len(left.union(right)) != len(indices) or len(left.union(right).difference(indices)) > 0:
```

Comment:
Just checking that we want to enforce that left union right == len(self)? (i.e. we probably don't want to ever construct a Datasets with elements we don't train or test over).

Reply:
I believe so, which is why I check that the possible indices (`indices = set(range(len(self)))`) are equal to `left.union(right)`